warp-3.3.24 has wrong bounds and doesn't compile with newest unix

unix-compat-0.7 was just released and this unblocked something, now the default warp build on ghc-9.6.1 errors.

This is with cabal-install 3.10 and the index-state is set to 2023-03-15T19:44:09Z.

```
% cabal repl -b warp
Resolving dependencies...
Build profile: -w ghc-9.6.1 -O1
In order, the following will be built (use -v for more details):
 - warp-3.3.24 (lib) (requires build)
 - fake-package-0 (lib) (first run)
Starting     warp-3.3.24 (lib)
Building     warp-3.3.24 (lib)

Failed to build warp-3.3.24.
Build log (
/home/janus/.cabal/logs/ghc-9.6.1/warp-3.3.24-b26c7fa72fe1d8ee610faa9eebd85286db66be49dac31832ab2c751f3f8a1f35.log
):
Configuring library for warp-3.3.24..
Preprocessing library for warp-3.3.24..
Building library for warp-3.3.24..
[ 1 of 34] Compiling Network.Wai.Handler.Warp.Date ( Network/Wai/Handler/Warp/Date.hs, dist/build/Network/Wai/Handler/Warp/Date.o, dist/build/Network/Wai/Handler/Warp/Date.dyn_o )
[ 2 of 34] Compiling Network.Wai.Handler.Warp.HashMap ( Network/Wai/Handler/Warp/HashMap.hs, dist/build/Network/Wai/Handler/Warp/HashMap.o, dist/build/Network/Wai/Handler/Warp/HashMap.dyn_o )
[ 3 of 34] Compiling Network.Wai.Handler.Warp.Imports ( Network/Wai/Handler/Warp/Imports.hs, dist/build/Network/Wai/Handler/Warp/Imports.o, dist/build/Network/Wai/Handler/Warp/Imports.dyn_o )
[ 4 of 34] Compiling Network.Wai.Handler.Warp.FileInfoCache ( Network/Wai/Handler/Warp/FileInfoCache.hs, dist/build/Network/Wai/Handler/Warp/FileInfoCache.o, dist/build/Network/Wai/Handler/Warp/FileInfoCache.dyn_o )
[ 5 of 34] Compiling Network.Wai.Handler.Warp.Counter ( Network/Wai/Handler/Warp/Counter.hs, dist/build/Network/Wai/Handler/Warp/Counter.o, dist/build/Network/Wai/Handler/Warp/Counter.dyn_o )
[ 6 of 34] Compiling Network.Wai.Handler.Warp.MultiMap ( Network/Wai/Handler/Warp/MultiMap.hs, dist/build/Network/Wai/Handler/Warp/MultiMap.o, dist/build/Network/Wai/Handler/Warp/MultiMap.dyn_o )
[ 7 of 34] Compiling Network.Wai.Handler.Warp.FdCache ( Network/Wai/Handler/Warp/FdCache.hs, dist/build/Network/Wai/Handler/Warp/FdCache.o, dist/build/Network/Wai/Handler/Warp/FdCache.dyn_o )

Network/Wai/Handler/Warp/FdCache.hs:70:11: error: [GHC-83865]
    • Couldn't match expected type: OpenFileFlags -> IO Fd
                  with actual type: IO Fd
    • The function ‘openFd’ is applied to four value arguments,
        but its type ‘FilePath -> OpenMode -> OpenFileFlags -> IO Fd’
        has only three
      In a stmt of a 'do' block:
        fd <- openFd
                path ReadOnly Nothing defaultFileFlags {nonBlock = False}
      In the expression:
        do fd <- openFd
                   path ReadOnly Nothing defaultFileFlags {nonBlock = False}
           setFileCloseOnExec fd
           return fd
   |
70 |     fd <- openFd path ReadOnly Nothing defaultFileFlags{nonBlock=False}
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Network/Wai/Handler/Warp/FdCache.hs:70:32: error: [GHC-83865]
    • Couldn't match expected type ‘OpenFileFlags’
                  with actual type ‘Maybe a0’
    • In the third argument of ‘openFd’, namely ‘Nothing’
      In a stmt of a 'do' block:
        fd <- openFd
                path ReadOnly Nothing defaultFileFlags {nonBlock = False}
      In the expression:
        do fd <- openFd
                   path ReadOnly Nothing defaultFileFlags {nonBlock = False}
           setFileCloseOnExec fd
           return fd
   |
70 |     fd <- openFd path ReadOnly Nothing defaultFileFlags{nonBlock=False}
```

--- a/warp/Network/Wai/Handler/Warp/FdCache.hs
+++ b/warp/Network/Wai/Handler/Warp/FdCache.hs
@@ -67,7 +67,11 @@ data FdEntry = FdEntry !Fd !MutableStatus
 
 openFile :: FilePath -> IO Fd
 openFile path = do
+#if MIN_VERSION_unix(2,8,0)
+    fd <- openFd path ReadOnly defaultFileFlags{nonBlock=False}
+#else
     fd <- openFd path ReadOnly Nothing defaultFileFlags{nonBlock=False}
+#endif
     setFileCloseOnExec fd
     return fd
 

